node/core/approval-voting/src/lib.rs

```diff
@@ -774,6 +793,162 @@ where
 	Ok(())
 }
 
+#[derive(Eq, PartialEq)]
+enum RecoveryState {
+	Pending,
+	Complete,
+}
+
+impl RecoveryState {
+	fn peek(&self) -> bool {
+		*self == RecoveryState::Complete
+	}
+
+	fn complete(&mut self) -> bool {
+		let complete = self.peek();
+		*self = RecoveryState::Complete;
+		complete
+	}
+}
+
+// Determines the set of assigned but unapproved candidates that require resumption of the
+// approval process by the local node. For each such, this method generates a `LaunchApproval`
+// action so that the approvals can continue to make progress after a restart.
+async fn handle_startup(
+	overlay_db: &mut OverlayedBackend<'_, impl Backend>,
+	state: &mut State,
+) -> SubsystemResult<Vec<(BlockNumber, Action)>> {
+	let all_candidates = overlay_db.load_all_candidates().map_err(|e| {
+		tracing::error!(target: LOG_TARGET, "Failed initial load of active candidates: {:?}", e);
+		e
+	})?;
+
+	let mut actions = Vec::new();
+	for candidate in all_candidates {
+		let mut assigned_but_unapproved_candidates = Vec::new();
+		let session = candidate.session;
+		let info = match state.session_info(session) {
+			Some(info) => info,
+			None => {
+				tracing::debug!(target: LOG_TARGET, "Missing session info for {:?}", session);
+				continue
+			},
+		};
+
+		// Compute the local validator indexes, if any, for this session.
+		let local_validators: Vec<_> = info
```

Comment:
It seems fairly heavy to do this for every candidate even though there are only a few possible sessions we might care about. I recommend caching the `local_validators` and `local_assigners` per-session in a `HashMap<SessionIndex, (Vec<ValidatorIndex>, Vec<ValidatorIndex>)>`